### DISCUSSION TITLE
Correct realizations of hexagonal lattice using MPB?

### DISCUSSION THREAD

**Discussion by knifelees3** [NanoComp/meep#2484] [General]
I want to realize the real hexagonal lattice using MPB, there will be six rods in the unit cell，
![image](https://user-images.githubusercontent.com/18129412/234008428-d8cca2ce-383c-4f36-8999-a302770dca45.png)

```
import math
import meep as mp
from meep import mpb
import numpy as np
import matplotlib.pyplot as plt

num_bands = 4

# 定义好扫描的k波矢
k_points = [mp.Vector3(),          # Gamma
            mp.Vector3(1/3,2/3),       # K
            mp.Vector3(0.5,0.5),  # M
            mp.Vector3()]          # Gamma

# 对k波矢进行插值
k_points = mp.interpolate(40, k_points)

#  units of nm
a=1



# 基本结构和材料性质
# 折射率设置
nslab = 2.48
slab = mp.Medium(index=nslab)

nair = 1
air = mp.Medium(index=nair)



s=0.25
geometry = [mp.Cylinder(0.1,center=mp.Vector3(-1*s,0), material=air),
            mp.Cylinder(0.1,center=mp.Vector3(-0.5*s,math.sqrt(3)/2*s), material=air),
            mp.Cylinder(0.1,center=mp.Vector3(0.5*s,math.sqrt(3)/2*s), material=air),
            mp.Cylinder(0.1,center=mp.Vector3(1*s,0), material=air),
           mp.Cylinder(0.1,center=mp.Vector3(0.5*s,-math.sqrt(3)/2*s), material=air),
            mp.Cylinder(0.1,center=mp.Vector3(-0.5*s1,-math.sqrt(3)/2*s), material=air)]



geometry_lattice = mp.Lattice(
    size=mp.Vector3(1, 1),
    basis1=mp.Vector3(0.5, np.sqrt(3) / 2),
    basis2=mp.Vector3(1,0),
)


resolution = 32

# mode solver设置
ms = mpb.ModeSolver(num_bands=num_bands,
                    k_points=k_points,
                    geometry=geometry,
                    geometry_lattice=geometry_lattice,
                    default_material=slab,
                    resolution=resolution)

ms.run_te()
te_freqs = ms.all_freqs
te_gaps = ms.gap_list
md = mpb.MPBData(rectify=True, periods=1, resolution=64)
eps = ms.get_epsilon()
converted_eps = md.convert(eps)

plt.imshow(converted_eps.T, interpolation='spline36', cmap='binary')
plt.axis('off')
plt.show()
```

However, the generated material distribution is not correct (I have used `rectify=True` to correct the distributions ), How can I resolve this?

![image](https://user-images.githubusercontent.com/18129412/234009136-0a009332-4ea3-494b-9439-0d25e7aa1081.png)

**Comment by stevengj** [comment]
Technically, this is a [honeycomb structure](https://en.wikipedia.org/wiki/Hexagonal_lattice) in a hexagonal lattice.   The MPB primitive unit cell is a *rhombus* (not a hexagon — that is a [Wigner–Seitz cell](https://en.wikipedia.org/wiki/Wigner%E2%80%93Seitz_cell), whereas in MPB you need to specify the primitive lattice vectors, which give a parallelogram/parallelopiped), and has *two* holes per primitive cell.

(You also need to understand [MPB's coordinate system](https://mpb.readthedocs.io/en/latest/Python_Tutorial/#a-few-words-on-units): the geometry centers are *not* Cartesian coordinates, they are in the basis of the lattice vectors.*)

You should start with the [triangular lattice (i.e. hexagonal lattice) tutorial](https://mpb.readthedocs.io/en/latest/Scheme_Tutorial/#bands-of-a-triangular-lattice) and add a *second* hole to the unit cell.

Actually, I just remembered that there is an [MPB example file for a honeycomb lattice](https://github.com/NanoComp/mpb/blob/a6da0e8ed2c20f04cf8a55c6f75668be6cdd6826/examples/honey-rods.ctl), albeit in Scheme — should be easy for you to copy from.

**Comment by stevengj** [comment]
PS. Even if you use a hexagonal Wigner–Seitz cell, there are still only two cylinders per primitive cell.  In your diagram, you drew your hexagons too big — the primitive Wigner–Seitz cell is a hexagon whose corners are the *centers* of the cylinders.  So, the cell contains only 1/3 of each cylinder, and 6/3 = 2 whole cylinders per Wigner–Seitz cell.

**Reply by knifelees3** [reply]
Thanks for your explanations, I can produce the corresponding honeycomb structure with nano rods (or holes), just by re-setting the geometry
```python
geometry = [mp.Cylinder(0.2,center=mp.Vector3(0,0), material=air),
            mp.Cylinder(0.1,center=mp.Vector3(0.25,math.sqrt(3)/4), material=air)]
```
![image](https://user-images.githubusercontent.com/18129412/234332613-b2b828a8-92f4-481e-b5ae-8f5f300514aa.png)


However, when using other shapes, such as square or hexagonal holes,  The desired shape is distorted.
As is shown below, 

![image](https://user-images.githubusercontent.com/18129412/234331043-140bfde7-d1db-4cd7-b306-be39529c47b5.png)

The following are the geometery settings for differet shapes
* square holes
```python
 geometry = [mp.Block(mp.Vector3(0.2,0.2),center=mp.Vector3(-0.25/2,-np.sqrt(3)/4/2), material=air),
            mp.Block(mp.Vector3(0.2,0.2),center=mp.Vector3(0.25/2,np.sqrt(3)/4/2), material=air)]
```
* hexagonal
```python
s1=0.1
s2=0.2
vertices1 = [mp.Vector3(-1*s1,0),
            mp.Vector3(-0.5*s1,math.sqrt(3)/2*s1),
            mp.Vector3(0.5*s1,math.sqrt(3)/2*s1),
            mp.Vector3(1*s1,0),
            mp.Vector3(0.5*s1,-math.sqrt(3)/2*s1),
            mp.Vector3(-0.5*s1,-math.sqrt(3)/2*s1)]

vertices2 = [mp.Vector3(-1*s2,0),
            mp.Vector3(-0.5*s2,math.sqrt(3)/2*s2),
            mp.Vector3(0.5*s2,math.sqrt(3)/2*s2),
            mp.Vector3(1*s2,0),
            mp.Vector3(0.5*s2,-math.sqrt(3)/2*s2),
            mp.Vector3(-0.5*s2,-math.sqrt(3)/2*s2)]
geometry = [mp.Prism(vertices1, height=mp.inf, center=mp.Vector3(0,0), material=air),
            
           mp.Prism(vertices2, height=mp.inf, center=mp.Vector3(0,np.sqrt(3)/4),material=air)]
```

The generated distribution is not I want
![image](https://user-images.githubusercontent.com/18129412/234329271-0b08da42-f71d-4901-a420-44b8d746cc74.png)
![image](https://user-images.githubusercontent.com/18129412/234329635-50ca4567-218a-4d92-a8f3-dfefcb48b48f.png)


Do I have to pre-warp the shape inside the cell in order to get the setup right? For example, a square is deformed and transformed into a corresponding rhombus, so that the final result is a square? I also don't understand why there is no problem with the round shape.

Or is it just a matter of MEEP display? Calculations are not affected?

**Reply by Ron-007-git** [reply]
Hello, did you find the solution of the above problem? Why do these hexagons look distorted?

**Reply by stevengj** [reply]
By default, all vectors in MPB are in the basis of the lattice vectors.  The default edges of a block are `e1 = (1,0)` and `e2 = (0,1)` — but again, these are in the lattice basis, which is *not* Cartesian in your case.

All you need to do is to call the `cartesian_to_lattice` function to convert your desired edge orientations to lattice coordinates, and set these as the `e1` and `e2` directions of the block.

**Reply by Ron-007-git** [reply]
Thank you for your reply. Now the blocks look fine but they seem too be rotated.  It would be really helpful if you could point out where I went wrong.

```py
geometry_lattice = mp.Lattice(
    size=mp.Vector3(1, 1),
    basis1=mp.Vector3(0.5, np.sqrt(3) / 2),
    basis2=mp.Vector3(1,0),
)



e1 = mp.cartesian_to_lattice(mp.Vector3(1, 0, 0), geometry_lattice)
e2 = mp.cartesian_to_lattice(mp.Vector3(0, 1, 0), geometry_lattice)
size=mp.Vector3(0.2, 0.2)

geometry = [mp.Block(size, e1, e2, center=mp.Vector3(-0.25/2, -np.sqrt(3)/4/2), material=air),
            mp.Block(size, e1, e2, center=mp.Vector3(0.25/2,np.sqrt(3)/4/2), material=air)
                     ]
```

![k](https://github.com/NanoComp/meep/assets/139942724/2cbda491-af29-4179-acf6-8bafa7c69c11)

**Reply by knifelees3** [reply]
> Hello, did you find the solution of the above problem? Why do these hexagons look distorted?

I didn' t find a solution

**Reply by stevengj** [reply]
Looks like you also want to change the `center` to use `cartesian_to_lattice`?
